aptos-move/framework/src/natives/aggregator_natives/helpers_v2.rs

```diff
@@ -1,48 +1,152 @@
+const DERIVED_STRING_SNAPSHOT_VALUE_FIELD_INDEX: usize = 0;
+const _DERIVED_STRING_SNAPSHOT_PADDING_FIELD_INDEX: usize = 1;
 
-pub(crate) fn aggregator_value_field_as_id(
-    value: u128,
-    resolver: &dyn DelayedFieldResolver,
-) -> PartialVMResult<DelayedFieldID> {
-    let value_u64 = u128_to_u64(value)?;
-    Ok(resolver.validate_and_convert_delayed_field_id(value_u64)?)
+macro_rules! get_value_impl {
+    ($func_name:ident, $idx:expr, $e:expr) => {
+        pub(crate) fn $func_name(struct_ref: &StructRef, ty: &Type) -> SafeNativeResult<u128> {
+            Ok(match ty {
+                Type::U128 => safely_get_struct_field_as!(struct_ref, $idx, u128),
+                Type::U64 => safely_get_struct_field_as!(struct_ref, $idx, u64) as u128,
+                _ => return Err(SafeNativeError::Abort { abort_code: $e }),
+            })
+        }
+    };
 }
 
-pub(crate) fn aggregator_snapshot_value_field_as_id(
-    value: u128,
-    resolver: &dyn DelayedFieldResolver,
-) -> PartialVMResult<DelayedFieldID> {
-    aggregator_value_field_as_id(value, resolver)
+macro_rules! get_value_as_id_impl {
+    ($func_name:ident, $idx:expr, $e:expr) => {
+        pub(crate) fn $func_name(
+            struct_ref: &StructRef,
+            ty: &Type,
+            resolver: &dyn DelayedFieldResolver,
+        ) -> SafeNativeResult<DelayedFieldID> {
+            let id = match ty {
+                Type::U64 | Type::U128 => {
+                    safely_get_struct_field_as!(struct_ref, $idx, DelayedFieldID)
+                },
+                _ => return Err(SafeNativeError::Abort { abort_code: $e }),
+            };
+
+            // Make sure we validate generated id is correct, i.e. it lies within the
+            // right bounds, etc. to catch bugs early.
+            resolver
+                .validate_delayed_field_id(&id)
+                .map_err(|e| SafeNativeError::InvariantViolation(PartialVMError::from(e)))?;
+            Ok(id)
+        }
+    };
 }
 
-pub(crate) fn derived_string_value_field_as_id(
-    value: Vec<u8>,
-    resolver: &dyn DelayedFieldResolver,
-) -> PartialVMResult<DelayedFieldID> {
-    let value_u64: u64 = from_utf8_bytes(value)?;
-    Ok(resolver.validate_and_convert_delayed_field_id(value_u64)?)
+get_value_impl!(
+    get_aggregator_value,
+    AGGREGATOR_VALUE_FIELD_INDEX,
+    EUNSUPPORTED_AGGREGATOR_TYPE
+);
+
+get_value_as_id_impl!(
+    get_aggregator_value_as_id,
+    AGGREGATOR_VALUE_FIELD_INDEX,
+    EUNSUPPORTED_AGGREGATOR_TYPE
+);
+
+pub(crate) fn set_aggregator_value(aggregator: &StructRef, value: Value) -> SafeNativeResult<()> {
+    aggregator
+        .borrow_field(AGGREGATOR_VALUE_FIELD_INDEX)
+        .map_err(SafeNativeError::InvariantViolation)?
+        .value_as::<Reference>()
+        .map_err(SafeNativeError::InvariantViolation)?
+        .write_ref(value)
+        .map_err(SafeNativeError::InvariantViolation)
 }
 
-pub(crate) fn set_aggregator_value_field(
+pub(crate) fn get_aggregator_max_value(
```

Comment:
I think this can be implemented with get_value_impl! macro also?

Reply:
indeed!